File: dotnet/src/LabelZoom.Sdk/Formats.cs
using System;
using System.Collections.Generic;

namespace LabelZoom.Sdk
{
    /// <summary>
    /// A format the LabelZoom API can convert <em>from</em>.
    /// </summary>
    /// <remarks>
    /// Deliberately a different type from <see cref="TargetFormat"/>. That is what makes
    /// <see cref="Epl"/>, <see cref="Tspl"/> and <see cref="Dpl"/> un-selectable as conversion
    /// targets: they are source-only on the server, and there is simply no
    /// <c>TargetFormat.Epl</c> to name.
    /// </remarks>
    public enum SourceFormat
    {
        /// <summary>Zebra Programming Language. Sent as <c>text/plain</c>.</summary>
        Zpl,

        /// <summary>Eltron Programming Language. Source-only. Sent as <c>text/plain</c>.</summary>
        Epl,

        /// <summary>TSC Printer Language. Source-only. Sent as <c>text/plain</c>.</summary>
        Tspl,

        /// <summary>Datamax Printer Language. Source-only. Sent as <c>text/plain</c>.</summary>
        Dpl,

        /// <summary>LabelZoom XML. Sent as <c>application/xml</c>.</summary>
        Xml,

        /// <summary>LabelZoom JSON. Sent as <c>application/json</c>.</summary>
        Json,

        /// <summary>PDF document. Sent as <c>application/pdf</c>.</summary>
        Pdf,

        /// <summary>PNG image. Sent as <c>image/png</c>.</summary>
        Png,

        /// <summary>BMP image. Sent as <c>image/bmp</c>.</summary>
        Bmp,

        /// <summary>GIF image. Sent as <c>image/gif</c>.</summary>
        Gif,

        /// <summary>JPEG image. Sent as <c>image/jpeg</c>.</summary>
        Jpeg,

        /// <summary>Alias for <see cref="Jpeg"/>; normalized to <c>jpeg</c> on the wire.</summary>
        Jpg,

        /// <summary>
        /// A URL, sent as the request body with <c>text/plain</c>. The <em>server</em> then fetches
        /// that URL and converts whatever it finds.
        /// </summary>
        /// <remarks>
        /// This hands a caller-supplied URL to a server-side fetch. Do not pass a URL derived from
        /// untrusted input without validating it first.
        /// </remarks>
        Url,
    }

    /// <summary>
    /// A format the LabelZoom API can convert <em>to</em>.
    /// </summary>
    /// <remarks>
    /// EPL, TSPL and DPL are intentionally absent — the server accepts them as sources only.
    /// </remarks>
    public enum TargetFormat
    {
        /// <summary>Zebra Programming Language. All labels are concatenated.</summary>
        Zpl,

        /// <summary>LabelZoom XML. First label only.</summary>
        Xml,

        /// <summary>LabelZoom JSON. First label only. Requires a paid license.</summary>
        Json,

        /// <summary>PDF document, one page per label.</summary>
        Pdf,

        /// <summary>PNG image. First label only.</summary>
        Png,

        /// <summary>BMP image. First label only.</summary>
        Bmp,

        /// <summary>GIF image. First label only.</summary>
        Gif,

        /// <summary>JPEG image. First label only.</summary>
        Jpeg,
    }

    /// <summary>Colour handling when rasterizing or tracing images.</summary>
    public enum ColorMode
    {
        /// <summary>Two-colour black and white.</summary>
        Bw,

        /// <summary>Greyscale. The server default.</summary>
        Grayscale,

        /// <summary>Full colour.</summary>
        Color,
    }

    /// <summary>How a source PDF is interpreted.</summary>
    public enum PdfConversionMode
    {
        /// <summary>Rasterize the page and trace it. The server default.</summary>
        Image,

        /// <summary>Read the PDF's native drawing operations.</summary>
        Native,
    }

    /// <summary>Image compression used when writing ZPL.</summary>
    public enum ZplImageCompression
    {
        /// <summary>Base-64 encoded, DEFLATE compressed. The server default.</summary>
        Z64,

        /// <summary>Run-length encoded hexadecimal.</summary>
        CompressedHex,
    }

    /// <summary>
    /// Wire tokens and media types for <see cref="SourceFormat"/> and <see cref="TargetFormat"/>.
    /// </summary>
    /// <remarks>
    /// This is the single place in the SDK that knows the format matrix. The superseded builder
    /// hierarchy spread the same knowledge across seven classes, and they drifted — one of them
    /// handled 2 of 12 sources, another returned the source content type as the target format.
    /// Keep it here.
    /// </remarks>
    internal static class Formats
    {
        private static readonly Dictionary<SourceFormat, string> SourceTokens =
            new Dictionary<SourceFormat, string>
            {
                [SourceFormat.Zpl] = "zpl",
                [SourceFormat.Epl] = "epl",
                [SourceFormat.Tspl] = "tspl",
                [SourceFormat.Dpl] = "dpl",
                [SourceFormat.Xml] = "xml",
                [SourceFormat.Json] = "json",
                [SourceFormat.Pdf] = "pdf",
                [SourceFormat.Png] = "png",
                [SourceFormat.Bmp] = "bmp",
                [SourceFormat.Gif] = "gif",
                // Rule A2: jpg normalizes to jpeg. One canonical wire spelling.
                [SourceFormat.Jpeg] = "jpeg",
                [SourceFormat.Jpg] = "jpeg",
                [SourceFormat.Url] = "url",
            };

        private static readonly Dictionary<SourceFormat, string> SourceMediaTypes =
            new Dictionary<SourceFormat, string>
            {
                [SourceFormat.Zpl] = "text/plain",
                [SourceFormat.Epl] = "text/plain",
                [SourceFormat.Tspl] = "text/plain",
                [SourceFormat.Dpl] = "text/plain",
                [SourceFormat.Xml] = "application/xml",
                [SourceFormat.Json] = "application/json",
                [SourceFormat.Pdf] = "application/pdf",
                [SourceFormat.Png] = "image/png",
                [SourceFormat.Bmp] = "image/bmp",
                [SourceFormat.Gif] = "image/gif",
                [SourceFormat.Jpeg] = "image/jpeg",
                [SourceFormat.Jpg] = "image/jpeg",
                [SourceFormat.Url] = "text/plain",
            };

        private static readonly Dictionary<TargetFormat, string> TargetTokens =
            new Dictionary<TargetFormat, string>
            {
                [TargetFormat.Zpl] = "zpl",
                [TargetFormat.Xml] = "xml",
                [TargetFormat.Json] = "json",
                [TargetFormat.Pdf] = "pdf",
                [TargetFormat.Png] = "png",
                [TargetFormat.Bmp] = "bmp",
                [TargetFormat.Gif] = "gif",
                [TargetFormat.Jpeg] = "jpeg",
            };

        internal static string ToWireToken(this SourceFormat format) =>
            SourceTokens.TryGetValue(format, out var token)
                ? token
                : throw new ArgumentOutOfRangeException(nameof(format), format, "Unknown source format.");

        internal static string ToMediaType(this SourceFormat format) =>
            SourceMediaTypes.TryGetValue(format, out var mediaType)
                ? mediaType
                : throw new ArgumentOutOfRangeException(nameof(format), format, "Unknown source format.");

        internal static string ToWireToken(this TargetFormat format) =>
            TargetTokens.TryGetValue(format, out var token)
                ? token
                : throw new ArgumentOutOfRangeException(nameof(format), format, "Unknown target format.");

        internal static string ToWireToken(this ColorMode mode) =>
            mode switch
            {
                ColorMode.Bw => "BW",
                ColorMode.Grayscale => "GRAYSCALE",
                ColorMode.Color => "COLOR",
                _ => throw new ArgumentOutOfRangeException(nameof(mode), mode, null),
            };

        internal static string ToWireToken(this PdfConversionMode mode) =>
            mode switch
            {
                PdfConversionMode.Image => "IMAGE",
                PdfConversionMode.Native => "NATIVE",
                _ => throw new ArgumentOutOfRangeException(nameof(mode), mode, null),
            };

        internal static string ToWireToken(this ZplImageCompression compression) =>
            compression switch
            {
                ZplImageCompression.Z64 => "Z64",
                ZplImageCompression.CompressedHex => "COMPRESSED_HEX",
                _ => throw new ArgumentOutOfRangeException(nameof(compression), compression, null),
            };
    }
}

File: rust/src/formats.rs
//! Source and target formats, and the one place a format's media type is written down.

use std::fmt;

/// The format of the document being converted.
///
/// A distinct type from [`TargetFormat`], and the two sets are not the same: [`Jpg`] and
/// [`Url`] are source-only. `Jpg` is an input spelling that normalizes to `jpeg` on the
/// wire, and `Url` is an instruction to the server to go fetch a document rather than a
/// format at all.
///
/// [`Jpg`]: SourceFormat::Jpg
/// [`Url`]: SourceFormat::Url
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum SourceFormat {
    /// Zebra Programming Language.
    Zpl,
    /// Eltron Programming Language.
    Epl,
    /// `TSC` Printer Language.
    Tspl,
    /// Datamax Programming Language.
    Dpl,
    /// The LabelZoom label model, as XML.
    Xml,
    /// The LabelZoom label model, as JSON.
    Json,
    /// A PDF document.
    Pdf,
    /// A PNG image.
    Png,
    /// A BMP image.
    Bmp,
    /// A GIF image.
    Gif,
    /// A JPEG image.
    Jpeg,
    /// An alias for [`SourceFormat::Jpeg`]; sent as `jpeg`.
    Jpg,
    /// Has the *server* fetch a URL and convert what it finds. The request body is the URL.
    ///
    /// Validate the URL first if it came from untrusted input: the fetch happens
    /// server-side, from LabelZoom's network, not from yours.
    Url,
}

/// The format to convert into.
///
/// There is no `TargetFormat::Url`: a URL is a fetch instruction, not an output format.
/// There is no `Jpg` either -- that spelling is source-side only.
///
/// The printer languages round-trip: `Epl`, `Tspl` and `Dpl` are targets as well as
/// sources, as of contract 1.1.0.
///
/// Marked `#[non_exhaustive]` on purpose. This set has already grown once -- the three
/// printer languages became targets when the printer-language writers shipped -- and
/// without it every such addition is a breaking change for any downstream `match`.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum TargetFormat {
    /// Zebra Programming Language.
    Zpl,
    /// Eltron Programming Language. Can inline raw binary (`GW`); prefer
    /// [`ConversionResult::bytes`](crate::ConversionResult::bytes) over `text()`.
    Epl,
    /// TSC Printer Language. Can inline raw binary (`BITMAP`); prefer
    /// [`ConversionResult::bytes`](crate::ConversionResult::bytes) over `text()`.
    Tspl,
    /// Datamax Programming Language.
    Dpl,
    /// The LabelZoom label model, as XML.
    Xml,
    /// The LabelZoom label model, as JSON.
    Json,
    /// A PDF document.
    Pdf,
    /// A PNG image.
    Png,
    /// A BMP image.
    Bmp,
    /// A GIF image.
    Gif,
    /// A JPEG image.
    Jpeg,
}

impl SourceFormat {
    /// Every accepted source, in the contract's order.
    pub const ALL: &'static [SourceFormat] = &[
        Self::Zpl,
        Self::Epl,
        Self::Tspl,
        Self::Dpl,
        Self::Xml,
        Self::Json,
        Self::Pdf,
        Self::Png,
        Self::Bmp,
        Self::Gif,
        Self::Jpeg,
        Self::Jpg,
        Self::Url,
    ];

    /// The path segment for this source. `Jpg` normalizes to `jpeg` (rule A2).
    pub fn wire_token(self) -> &'static str {
        match self {
            Self::Jpg => "jpeg",
            other => other.name(),
        }
    }

    /// The request `Content-Type` for this source.
    ///
    /// The format metadata table, defined once and never inlined at a call site. The
    /// superseded .NET design had a builder type per format, each independently knowing
    /// this mapping, and they drifted -- one of them emitted the source type as the target.
    pub fn media_type(self) -> &'static str {
        match self {
            // The body for Url is the URL itself.
            Self::Zpl | Self::Epl | Self::Tspl | Self::Dpl | Self::Url => "text/plain",
            Self::Xml => "application/xml",
            Self::Json => "application/json",
            Self::Pdf => "application/pdf",
            Self::Png => "image/png",
            Self::Bmp => "image/bmp",
            Self::Gif => "image/gif",
            Self::Jpeg | Self::Jpg => "image/jpeg",
        }
    }

    /// The lowercase name of this format, before `jpg` normalization.
    pub fn name(self) -> &'static str {
        match self {
            Self::Zpl => "zpl",
            Self::Epl => "epl",
            Self::Tspl => "tspl",
            Self::Dpl => "dpl",
            Self::Xml => "xml",
            Self::Json => "json",
            Self::Pdf => "pdf",
            Self::Png => "png",
            Self::Bmp => "bmp",
            Self::Gif => "gif",
            Self::Jpeg => "jpeg",
            Self::Jpg => "jpg",
            Self::Url => "url",
        }
    }

    /// Parses a wire token, for deserializing configuration.
    pub fn from_name(name: &str) -> Option<Self> {
        Self::ALL
            .iter()
            .copied()
            .find(|format| format.name() == name.to_ascii_lowercase())
    }
}

impl TargetFormat {
    /// Every accepted target, in the contract's order.
    pub const ALL: &'static [TargetFormat] = &[
        Self::Zpl,
        Self::Epl,
        Self::Tspl,
        Self::Dpl,
        Self::Xml,
        Self::Json,
        Self::Pdf,
        Self::Png,
        Self::Bmp,
        Self::Gif,
        Self::Jpeg,
    ];

    /// The path segment for this target. Targets need no normalization.
    pub fn wire_token(self) -> &'static str {
        match self {
            Self::Zpl => "zpl",
            Self::Epl => "epl",
            Self::Tspl => "tspl",
            Self::Dpl => "dpl",
            Self::Xml => "xml",
            Self::Json => "json",
            Self::Pdf => "pdf",
            Self::Png => "png",
            Self::Bmp => "bmp",
            Self::Gif => "gif",
            Self::Jpeg => "jpeg",
        }
    }

    /// Parses a wire token, for deserializing configuration.
    pub fn from_name(name: &str) -> Option<Self> {
        Self::ALL
            .iter()
            .copied()
            .find(|format| format.wire_token() == name.to_ascii_lowercase())
    }
}

impl fmt::Display for SourceFormat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.name())
    }
}

impl fmt::Display for TargetFormat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.wire_token())
    }
}
